delete SNAT2NIP if pod.node == egressNodeServingPod

tssurya opened the pull request:
We should be deleting the snat towards nodeIP before
adding snat towards egressIP only from the node
where the pod lives if that same node is serving
as the egressIPnode for the pod.

Signed-off-by: Surya Seetharaman <suryaseetharaman.9@gmail.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/ovn-org/ovn-kubernetes/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Trivial changes are exempt from following this template.
If your change is non-trivial, please provide the following information:
-->

**- What this PR does and why is it needed**
<!--
A summary of the changes within this pull request and some context
as to why they were made
-->

**- Special notes for reviewers**
<!--
What exactly did you change - you may also defer to information
contained in commit messages. At a bare minimum it's worth highlighting
which areas of the code were changed as it's easier to assign reviewers
-->


**- How to verify it**
<!--
Did you include unit tests? or end-to-end tests?
How can I manually verify that this patch achieves its objective
-->


**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

tssurya commented:
/assign @kyrtapz

tssurya commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -991,7 +988,7 @@ func (oc *Controller) addPodEgressIPAssignments(name string, statusAssignments [
 	}
 	for _, status := range remainingAssignments {
-		klog.V(5).Infof("Adding pod egress IP status: %v for EgressIP: %s and pod: %s/%s", status, name, pod.Name, pod.Namespace)
+		klog.Infof("Adding pod egress IP status: %v for EgressIP: %s and pod: %s/%s", status, name, pod.Name, pod.Namespace)
```
so why is it that this useful log is not printed by default and its log 5? :D Is it because it will generate too much noise when egressIPs move around nodes or pods are deleted/created? @flavio-fernandes @kyrtapz 
I can remove this change from this PR, but for me this line was very useful to see what is being assigned to what pod, but perhaps with the latest libovsdb changes we are anyways able to see each transact happening so maybe we are covered.

kyrtapz replied:
> so why is it that this useful log is not printed by default and its log 5? :D Is it because it will generate too much noise when egressIPs move around nodes or pods are deleted/created?

I think the log level was intentional as it can result in a lot of noise. It is printed every time there is a egress IP configured for a pod. With big clusters and ovn-k restarts it could potential result in a **lot** of log messages.

flavio-fernandes replied:
yeah, I would like this log to be there, but Alex convinced me it was too chatty to be useful.

tssurya commented:
@kyrtapz pointed out a very valid point: "What happens when the egressIP shifts to another node?" :)
Ans: We'll need to add back the snat towards nodeIP.
/hold PR reviews while I fix this. I also want to add an e2e test so that in future if we decide to remove the 101's all together under the presumption that pod2diff-node traffic is external then we can catch that.

tssurya commented:
> @kyrtapz pointed out a very valid point: "What happens when the egressIP shifts to another node?" :) Ans: We'll need to add back the snat towards nodeIP. 

actually I think we are good. I tested this:
```
$ oc get pods -n foo -owide
NAME                   READY   STATUS    RESTARTS   AGE     IP           NODE          NOMINATED NODE   READINESS GATES
client-on-ovn-worker   1/1     Running   0          7m42s   10.244.1.3   ovn-worker2   <none>           <none>
$ oc describe eip egressips-prod
Spec:
  Egress I Ps:
    172.18.0.9
  Namespace Selector:
    Match Labels:
      Env:  prod
Status:
  Items:
    Egress IP:  172.18.0.9
    Node:       ovn-worker2
Events:         <none>
# ovn-nbctl lr-nat-list GR_ovn-worker2
TYPE             EXTERNAL_IP        EXTERNAL_PORT    LOGICAL_IP            EXTERNAL_MAC         LOGICAL_PORT
snat             172.18.0.9                          10.244.1.3
```
Now we change the egressip towards ovn-worker:
```
$ oc describe eip egressips-prod
Spec:
  Egress I Ps:
    172.18.0.9
  Namespace Selector:
    Match Labels:
      Env:  prod
Status:
  Items:
    Egress IP:  172.18.0.9
    Node:       ovn-worker
Events:         <none>
sh-5.1# ovn-nbctl lr-nat-list GR_ovn-worker2
TYPE             EXTERNAL_IP        EXTERNAL_PORT    LOGICAL_IP            EXTERNAL_MAC         LOGICAL_PORT
snat             172.18.0.3                          10.244.1.3
sh-5.1# ovn-nbctl lr-nat-list GR_ovn-worker
TYPE             EXTERNAL_IP        EXTERNAL_PORT    LOGICAL_IP            EXTERNAL_MAC         LOGICAL_PORT
snat             172.18.0.9                          10.244.1.3
```
It works out of the box and reason is because when we re-assign egressIPs, we surely call the `addPerPodGRSNAT` either from `deleteEgressIPAssignments` OR `deletePodEgressIPAssignments` (not sure which one) and so adds back the SNAT towards the nodeIP on the GR-node where the pod lives.

flavio-fernandes commented on go-controller/pkg/ovn/egressip.go:
```diff
@@ -1843,9 +1843,10 @@ func (e *egressIPController) addPerPodGRSNAT(podNamespace, podName string, podIP
-	if config.Gateway.DisableSNATMultipleGWs {
-		// remove snats to->nodeIP (from the node where pod exists) for these podIPs before adding the snat to->egressIP
+func (e *egressIPController) deletePerPodGRSNAT(pod *kapi.Pod, podIPs []*net.IPNet, status egressipv1.EgressIPStatusItem) error {
+	if config.Gateway.DisableSNATMultipleGWs && status.Node == pod.Spec.NodeName {
```
@tssurya do you think it would be good to have an info log here to indicate we did not delete because 
`status.Node != pod.Spec.NodeName` ?

```
if config.Gateway.DisableSNATMultipleGWs {
   if status.Node != pod.Spec.NodeName {
       log ( "will not deletePerPodGRSNAT because pod is in the egress node ans Surya fixed this now!" )
   } else {
     ...
```

tssurya replied:
good point, might create some noise though maybe log5...

flavio-fernandes reviewed (commented):
some comments.
BUT, this LGTM!

tssurya commented:
The one thing I am concerned about is what Alex said in commit message: https://github.com/ovn-org/ovn-kubernetes/pull/2734/commits/9353e949c42ea6db4da7be824fd3043af18738f6 ; he does have a point though that with every update I'm possibly trying a SNAT delete/create because rn I changed the position from where we call the `deletePerPodGRSNAT` because he had split that call to be done only once for a given pod... But I'm not sure how else we can make sure our setup is correct always because if egressIP change happens, then depending on that, we do need to delete/add the SNAT. So I guess for now, correctness before performace...Hopefully once we solve that inconsistency around service/host-net traffic for egressIPs, we'll get more clarity.

tssurya commented:
The lgw CI jobs are failing on my new e2e, not sure what's wrong... taking a look.

tssurya commented:
> The lgw CI jobs are failing on my new e2e, not sure what's wrong... taking a look.

there is inconsistency wrt the gateway modes as well, lgw behaves different from sgw...will try to capture this in the test.

tssurya commented:
actually the egress fail over scenario has some delays which is causing this CI flake:
```
STEP: 0. Add the "k8s.ovn.org/egress-assignable" label to egress1Node node
Apr 25 22:23:04.871: INFO: Added egress-assignable label to node ovn-worker
STEP: 1. Creating host-networked pod, on non-egress node to act as "another node"
Apr 25 22:23:10.891: INFO: Created pod ovn-control-plane-host-net-pod on node ovn-control-plane
STEP: 2. Create an EgressIP object with one egress IP defined
Apr 25 22:23:10.897: INFO: Create the EgressIP configuration
Apr 25 22:23:10.897: INFO: Running '/usr/local/bin/kubectl --server=https://127.0.0.1:35141/ --kubeconfig=/home/runner/ovn.conf --namespace=default create -f egressip.yaml'
Apr 25 22:23:11.035: INFO: stderr: ""
Apr 25 22:23:11.035: INFO: stdout: "egressip.k8s.ovn.org/egressip created\n"
STEP: 3. Check that the status is of length one and that it is assigned to egress1Node
Apr 25 22:23:11.035: INFO: Running '/usr/local/bin/kubectl --server=https://127.0.0.1:35141/ --kubeconfig=/home/runner/ovn.conf --namespace=default get eip -o json'
Apr 25 22:23:11.099: INFO: stderr: ""
Apr 25 22:23:11.099: INFO: stdout: "{\n    \"apiVersion\": \"v1\",\n    \"items\": [\n        {\n            \"apiVersion\": \"k8s.ovn.org/v1\",\n            \"kind\": \"EgressIP\",\n            \"metadata\": {\n                \"creationTimestamp\": \"2022-04-25T22:23:11Z\",\n                \"generation\": 2,\n                \"name\": \"egressip\",\n                \"resourceVersion\": \"5850\",\n                \"uid\": \"9d261b19-83cc-4356-b9b9-db256da6ff7b\"\n            },\n            \"spec\": {\n                \"egressIPs\": [\n                    \"172.18.1.3\"\n                ],\n                \"namespaceSelector\": {\n                    \"matchLabels\": {\n                        \"name\": \"egressip-9267\"\n                    }\n                },\n                \"podSelector\": {\n                    \"matchLabels\": {\n                        \"wants\": \"egress\"\n                    }\n                }\n            },\n            \"status\": {\n                \"items\": [\n                    {\n                        \"egressIP\": \"172.18.1.3\",\n                        \"node\": \"ovn-worker\"\n                    }\n                ]\n            }\n        }\n    ],\n    \"kind\": \"List\",\n    \"metadata\": {\n        \"resourceVersion\": \"\",\n        \"selfLink\": \"\"\n    }\n}\n"

Original node: ovn-worker ^

STEP: 7. Add the "k8s.ovn.org/egress-assignable" label to egress2Node
Apr 25 22:23:15.650: INFO: Added egress-assignable label to node ovn-control-plane
STEP: 8. Remove the "k8s.ovn.org/egress-assignable" label from egress1Node
STEP: removing the label k8s.ovn.org/egress-assignable off the node ovn-worker
STEP: verifying the node doesn't have the label k8s.ovn.org/egress-assignable
STEP: 9. Check that the status is of length one and that it is assigned to egress2Node
Apr 25 22:23:15.678: INFO: Running '/usr/local/bin/kubectl --server=https://127.0.0.1:35141/ --kubeconfig=/home/runner/ovn.conf --namespace=default get eip -o json'
Apr 25 22:23:15.795: INFO: stderr: ""
Apr 25 22:23:15.795: INFO: stdout: "{\n    \"apiVersion\": \"v1\",\n    \"items\": [\n        {\n            \"apiVersion\": \"k8s.ovn.org/v1\",\n            \"kind\": \"EgressIP\",\n            \"metadata\": {\n                \"creationTimestamp\": \"2022-04-25T22:23:11Z\",\n                \"generation\": 2,\n                \"name\": \"egressip\",\n                \"resourceVersion\": \"5850\",\n                \"uid\": \"9d261b19-83cc-4356-b9b9-db256da6ff7b\"\n            },\n            \"spec\": {\n                \"egressIPs\": [\n                    \"172.18.1.3\"\n                ],\n                \"namespaceSelector\": {\n                    \"matchLabels\": {\n                        \"name\": \"egressip-9267\"\n                    }\n                },\n                \"podSelector\": {\n                    \"matchLabels\": {\n                        \"wants\": \"egress\"\n                    }\n                }\n            },\n            \"status\": {\n                \"items\": [\n                    {\n                        \"egressIP\": \"172.18.1.3\",\n                        \"node\": \"ovn-worker\"\n                    }\n                ]\n            }\n        }\n    ],\n    \"kind\": \"List\",\n    \"metadata\": {\n        \"resourceVersion\": \"\",\n        \"selfLink\": \"\"\n    }\n}\n"
Apr 25 22:23:15.795: FAIL: Step 2. Check that the status is of length one and that it is assigned to egress2Node, failed
```
even if we removed it from ovn-worker and added it to ovn-control-plane, the eip status takes time to reflect ?

tssurya commented:
```

JUnit report was created: /home/runner/work/ovn-kubernetes/ovn-kubernetes/go-controller/_artifacts/junit-pkg_ovn.xml

Ran 142 of 142 Specs in 196.845 seconds
SUCCESS! -- 142 Passed | 0 Failed | 0 Pending | 0 Skipped
--- PASS: TestClusterNode (197.07s)
FAIL
coverage: 63.5% of statements
FAIL	github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn	218.936s
FAIL
make: *** [Makefile:52: check] Error 1
Error: Process completed with exit code 2.
```
not sure why this errors out when all the unit tests have passed...

tssurya commented:
/retest-failed

andreaskaris commented on test/e2e/e2e.go:
```diff
@@ -1160,6 +1176,182 @@ spec:
+		// There is sometimes a slight delay for the EIP fail over to happen,
+		// so let's use the pollimmediate struct to check if eventually egress2Node becomes the egress node
+		err = wait.PollImmediate(retryInterval, retryTimeout, func() (bool, error) {return statuses[0].Node == egress2Node.name, nil})
+		framework.ExpectNoError(err, "Step 2. Check that the status is of length one and that it is assigned to egress2Node, failed: %v", err)
```
"Step 9."

tssurya replied:
done

andreaskaris commented on test/e2e/e2e.go:
```diff
@@ -1160,6 +1176,182 @@ spec:
+
+		ginkgo.By("10. Check connectivity from pod to an external \"node\" and verify that the srcIP is the expected egressIP")
+		err = wait.PollImmediate(retryInterval, retryTimeout, targetExternalContainerAndTest(targetNode, pod1Name, podNamespace.Name, true, []string{egressIP1.String()}))
+		framework.ExpectNoError(err, "Step 9. Check connectivity from pod to an external \"node\" and verify that the srcIP is the expected egressIP, failed, err: %v", err)
```
"Step 10."

tssurya replied:
done

andreaskaris commented on test/e2e/e2e.go:
```diff
@@ -1160,6 +1176,182 @@ spec:
+		pod2.Labels = podEgressLabel
+		updatePod(f, pod2)
+
+		ginkgo.By("14. Check connectivity from second pod to external node and verify that the srcIP is the expected egressIP")
```
15.

tssurya replied:
done.

andreaskaris commented on test/e2e/e2e.go:
```diff
@@ -1160,6 +1176,182 @@ spec:
+
+		ginkgo.By("14. Check connectivity from second pod to external node and verify that the srcIP is the expected egressIP")
+		err = wait.PollImmediate(retryInterval, retryTimeout, targetExternalContainerAndTest(targetNode, pod2Name, podNamespace.Name, true, []string{egressIP1.String()}))
+		framework.ExpectNoError(err, "Step 14. Check connectivity from second pod to external node and verify that the srcIP is the expected egressIP, failed: %v", err)
```
15.

tssurya replied:
done.

andreaskaris commented on test/e2e/e2e.go:
```diff
@@ -1160,6 +1176,182 @@ spec:
+		err = wait.PollImmediate(retryInterval, retryTimeout, targetExternalContainerAndTest(targetNode, pod2Name, podNamespace.Name, true, []string{egressIP1.String()}))
+		framework.ExpectNoError(err, "Step 14. Check connectivity from second pod to external node and verify that the srcIP is the expected egressIP, failed: %v", err)
+
+		ginkgo.By("15. Check connectivity from second pod to another node and verify that the srcIP is the expected nodeIP (this verifies SNAT's towards nodeIP are not deleted unless node is egressNode)")
```
16.

tssurya replied:
done.

andreaskaris commented on test/e2e/e2e.go:
```diff
@@ -1160,6 +1176,182 @@ spec:
+
+		ginkgo.By("15. Check connectivity from second pod to another node and verify that the srcIP is the expected nodeIP (this verifies SNAT's towards nodeIP are not deleted unless node is egressNode)")
+		err = wait.PollImmediate(retryInterval, retryTimeout, targetExternalContainerAndTest(hostNetPod, pod2Name, podNamespace.Name, true, []string{egressNodeIP.String()}))
+		framework.ExpectNoError(err, "Step 15. Check connectivity from second pod to another node and verify that the srcIP is the expected nodeIP (this verifies SNAT's towards nodeIP are not deleted unless node is egressNode), failed: %v", err)
```
16.

tssurya replied:
done.

andreaskaris reviewed (changes requested):
So, I reviewed the obvious and this here all looks good with a few minor issues. However, please see my next comment for another big issue - I don't know if that is related to your change or if we should address this in a follow-up fix.

=============================================================

Review steps:

### upstream/master

Setup a disable snat multiple gws environment, in shared gateway mode:
~~~
git checkout upstream/master
contrib/kind.sh --disable-snat-multiple-gws
~~~

Apply the baseline setup from [0]

That yields:
~~~
[root@ovnkubernetes ~]# oc get pods -o wide -A | grep egress
egressip-source      egressip-source-deployment-56d66c78dd-cwb5l   1/1     Running   0          74s     10.244.1.6   ovn-worker          <none>           <none>
egressip-source      egressip-source-deployment-56d66c78dd-ln6n5   1/1     Running   0          74s     10.244.0.3   ovn-worker2         <none>           <none>
egressip-target      egressip-target-deployment-cgd92              1/1     Running   0          84s     172.18.0.2   ovn-control-plane   <none>           <none>
~~~

Test without egressip:
~~~
[root@ovnkubernetes ~]# oc exec -n $SOURCE_NAMESPACE egressip-source-deployment-56d66c78dd-cwb5l -- curl -s  172.18.0.2:$TARGET_PORT/clientip
172.18.0.3:48208[root@ovnkubernetes ~]# oc exec -n $SOURCE_NAMESPACE egressip-source-deployment-56d66c78dd-ln6n5.18.0.2:$TARGET_PORT/clientip
[root@ovnkubernetes ~]# oc exec -n $SOURCE_NAMESPACE egressip-source-deployment-56d66c78dd-ln6n5 -- curl -s  172.18.0.2:$TARGET_PORT/clientip
172.18.0.4:44484[root@ovnkubernetes ~]# 
~~~

Applying egressip:
~~~
file=$(mktemp)
cat <<EOF >| ${file}
apiVersion: k8s.ovn.org/v1
kind: EgressIP
metadata:
  name: ${EGRESSIP_NAME}
spec:
  egressIPs: [ "172.18.0.10" ]
  namespaceSelector:
    matchLabels:
      env: ${SOURCE_NAMESPACE}
EOF
oc apply -f $file
~~~

Failure can be reproduced:
~~~
[root@ovnkubernetes ~]# oc get  egressip 
NAME              EGRESSIPS     ASSIGNED NODE   ASSIGNED EGRESSIPS
egressip-source   172.18.0.10   ovn-worker2     172.18.0.10
[root@ovnkubernetes ~]# oc get pods -A -o wide | grep egressip
egressip-source      egressip-source-deployment-56d66c78dd-cwb5l   1/1     Running   0          5m35s   10.244.1.6   ovn-worker          <none>           <none>
egressip-source      egressip-source-deployment-56d66c78dd-ln6n5   1/1     Running   0          5m35s   10.244.0.3   ovn-worker2         <none>           <none>
egressip-target      egressip-target-deployment-cgd92              1/1     Running   0          5m45s   172.18.0.2   ovn-control-plane   <none>           <none>
[root@ovnkubernetes ~]# oc exec -n $SOURCE_NAMESPACE egressip-source-deployment-56d66c78dd-cwb5l -- curl -s  172.18.0.2:$TARGET_PORT/clientip
^C
[root@ovnkubernetes ~]# oc exec -n $SOURCE_NAMESPACE egressip-source-deployment-56d66c78dd-ln6n5 -- curl -s  172.18.0.2:$TARGET_PORT/clientip
172.18.0.10:45714[root@ovnkubernetes ~]# 
~~~

Missing SNAT entry on GR_ovn-worker for pod 10.244.1.6:
~~~
root@ovnkubernetes ~]# oc rsh -n ovn-kubernetes $(oc get pods -o name -n ovn-kubernetes -l name=ovnkube-master)
sh-5.1# ovn-nbctl lr-list
94c9bb0b-77d1-4144-bcb5-59ce6e0f08ac (GR_ovn-control-plane)
aa3cdc00-8829-43a3-9833-1f5e85fe1c84 (GR_ovn-worker)
69368d1e-8025-409c-86f4-1eb4990ac831 (GR_ovn-worker2)
c27667fa-925e-4057-97d6-45e093b478a2 (ovn_cluster_router)
sh-5.1# ovn-nbctl lr-nat-list GR_ovn-worker2    
TYPE             EXTERNAL_IP        EXTERNAL_PORT    LOGICAL_IP            EXTERNAL_MAC         LOGICAL_PORT
snat             172.18.0.10                         10.244.1.6
snat             172.18.0.10                         10.244.0.3
sh-5.1# ovn-nbctl lr-nat-list GR_ovn-worker 
TYPE             EXTERNAL_IP        EXTERNAL_PORT    LOGICAL_IP            EXTERNAL_MAC         LOGICAL_PORT
snat             172.18.0.3                          10.244.1.5
snat             172.18.0.3                          10.244.1.4
snat             172.18.0.3                          10.244.1.3
~~~

Roll back the setup:
~~~
oc delete egressip egressip-source
oc delete ns egressip-source ; oc delete ns egressip-target
oc adm taint node ovn-control-plane ${TARGET_TAINT}=true:NoExecute- --overwrite
oc adm taint node ovn-worker ${TARGET_TAINT}=true:NoExecute- --overwrite
oc adm taint node ovn-worker2 ${TARGET_TAINT}=true:NoExecute- --overwrite
oc label node ovn-control-plane ${TARGET_LABEL}-
oc label node ovn-worker ${SOURCE_LABEL}-
oc label node ovn-worker2 ${SOURCE_LABEL}-
oc label node ovn-control-plane ${EGRESS_ASSIGNABLE_LABEL}- --overwrite
oc label node ovn-worker ${EGRESS_ASSIGNABLE_LABEL}- --overwrite
oc label node ovn-worker2  ${EGRESS_ASSIGNABLE_LABEL}- --overwrite
~~~

Checkout the patch:
~~~
git checkout tssurya/fix-snats-gr
~~~

Run the E2E:
~~~
[root@ovnkubernetes ~]# cd development/ovn-kubernetes/test/e2e/
[root@ovnkubernetes e2e]# go test -v . --count=1 -ginkgo.focus "Should validate the egress IP SNAT functionality against host-networked pods" -ginkgo.v  
~~~

This fails at step 11):
~~~
  Apr 27 09:42:53.021: Step 11. Check connectivity from pod to another node and verify that the srcIP is the expected nodeIP (known inconsistency compared to step 6), failed: timed out waiting for the condition
  Unexpected error:
      <*errors.errorString | 0xc00033b140>: {
          s: "timed out waiting for the condition",
      }
      timed out waiting for the condition
  occurred
~~~

Make sure that the tests cleans up all resources:
~~~
[root@ovnkubernetes e2e]# oc get nodes -o yaml | grep assign
[root@ovnkubernetes e2e]# oc get egressip
No resources found
[root@ovnkubernetes e2e]# oc get pods -A -o wide | grep egress
[root@ovnkubernetes e2e]# 
~~~

### With patch

Checkout the patch:
~~~
git checkout tssurya/fix-snats-gr
contrib/kind.sh --disable-snat-multiple-gws
~~~

Apply the baseline setup from [0]

That yields:
~~~
[root@ovnkubernetes ~]# oc get pods -o wide -A | grep egress
egressip-source      egressip-source-deployment-56d66c78dd-hpzgk   0/1     ContainerCreating   0          16s     <none>       ovn-worker2         <none>           <none>
egressip-source      egressip-source-deployment-56d66c78dd-kg9gn   0/1     ContainerCreating   0          16s     <none>       ovn-worker          <none>           <none>
egressip-target      egressip-target-deployment-dj5zw              1/1     Running             0          24s     172.18.0.2   ovn-control-plane   <none>           <none>
[root@ovnkubernetes ~]# 
~~~

Test without egressip:
~~~
[root@ovnkubernetes ~]# oc exec -n $SOURCE_NAMESPACE egressip-source-deployment-56d66c78dd-hpzgk -- curl -s  172.18.0.2:$TARGET_PORT/clientip
172.18.0.4:50964[root@ovnkubernetes ~]# 
[root@ovnkubernetes ~]# oc exec -n $SOURCE_NAMESPACE egressip-source-deployment-56d66c78dd-kg9gn -- curl -s  172.18.0.2:$TARGET_PORT/clientip
172.18.0.3:33628[root@ovnkubernetes ~]# 
~~~

Applying egressip:
~~~
file=$(mktemp)
cat <<EOF >| ${file}
apiVersion: k8s.ovn.org/v1
kind: EgressIP
metadata:
  name: ${EGRESSIP_NAME}
spec:
  egressIPs: [ "172.18.0.10" ]
  namespaceSelector:
    matchLabels:
      env: ${SOURCE_NAMESPACE}
EOF
oc apply -f $file
~~~

Failure *cannot* be reproduced:
~~~
[root@ovnkubernetes ~]#  oc get  egressip 
NAME              EGRESSIPS     ASSIGNED NODE   ASSIGNED EGRESSIPS
egressip-source   172.18.0.10   ovn-worker2     172.18.0.10
[root@ovnkubernetes ~]# oc get nodes -o wide
NAME                STATUS   ROLES                                       AGE     VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE       KERNEL-VERSION          CONTAINER-RUNTIME
ovn-control-plane   Ready    control-plane,egressip-test-target,master   8m10s   v1.23.3   172.18.0.2    <none>        Ubuntu 21.10   4.18.0-373.el8.x86_64   containerd://1.5.9
ovn-worker          Ready    egressip-test-source                        7m32s   v1.23.3   172.18.0.3    <none>        Ubuntu 21.10   4.18.0-373.el8.x86_64   containerd://1.5.9
ovn-worker2         Ready    egressip-test-source                        7m32s   v1.23.3   172.18.0.4    <none>        Ubuntu 21.10   4.18.0-373.el8.x86_64   containerd://1.5.9
[root@ovnkubernetes ~]# oc get pods -A -o wide | grep egressip
egressip-source      egressip-source-deployment-56d66c78dd-hpzgk   1/1     Running   0          107s    10.244.2.3   ovn-worker2         <none>           <none>
egressip-source      egressip-source-deployment-56d66c78dd-kg9gn   1/1     Running   0          107s    10.244.1.6   ovn-worker          <none>           <none>
egressip-target      egressip-target-deployment-dj5zw              1/1     Running   0          115s    172.18.0.2   ovn-control-plane   <none>           <none>
[root@ovnkubernetes ~]# oc exec -n $SOURCE_NAMESPACE egressip-source-deployment-56d66c78dd-hpzgk -- curl -s  172.18.0.2:$TARGET_PORT/clientip
172.18.0.10:51640[root@ovnkubernetes ~]# 
[root@ovnkubernetes ~]# oc exec -n $SOURCE_NAMESPACE egressip-source-deployment-56d66c78dd-kg9gn -- curl -s  172.18.0.2:$TARGET_PORT/clientip
172.18.0.3:34222[root@ovnkubernetes ~]# 
~~~

SNAT entry on GR_ovn-worker for pod 10.244.1.6 exists:
~~~
[root@ovnkubernetes ~]#  oc rsh -n ovn-kubernetes $(oc get pods -o name -n ovn-kubernetes -l name=ovnkube-master)
Defaulting container name to ovn-northd.
Use 'oc describe pod/ovnkube-master-85d5b4f6b9-5sjsq -n ovn-kubernetes' to see all of the containers in this pod.
sh-5.1# ovn-nbctl lr-list
48b6b3b3-3b95-478a-86b2-878106a2a9a9 (GR_ovn-control-plane)
67407ed8-8cb8-44fc-be18-56f6273f337d (GR_ovn-worker)
02a9ea8f-3770-42df-869a-4e3b6a3654b0 (GR_ovn-worker2)
b153ecb0-dc3d-4a46-b8cd-9796210c3cf0 (ovn_cluster_router)
sh-5.1#  ovn-nbctl lr-nat-list GR_ovn-worker2 
TYPE             EXTERNAL_IP        EXTERNAL_PORT    LOGICAL_IP            EXTERNAL_MAC         LOGICAL_PORT
snat             172.18.0.10                         10.244.1.6
snat             172.18.0.10                         10.244.2.3
sh-5.1#  ovn-nbctl lr-nat-list GR_ovn-worker 
TYPE             EXTERNAL_IP        EXTERNAL_PORT    LOGICAL_IP            EXTERNAL_MAC         LOGICAL_PORT
snat             172.18.0.3                          10.244.1.3
snat             172.18.0.3                          10.244.1.6
snat             172.18.0.3                          10.244.1.5
snat             172.18.0.3                          10.244.1.4
~~~

Roll back the setup:
~~~
oc delete egressip egressip-source
oc delete ns egressip-source ; oc delete ns egressip-target
oc adm taint node ovn-control-plane ${TARGET_TAINT}=true:NoExecute- --overwrite
oc adm taint node ovn-worker ${TARGET_TAINT}=true:NoExecute- --overwrite
oc adm taint node ovn-worker2 ${TARGET_TAINT}=true:NoExecute- --overwrite
oc label node ovn-control-plane ${TARGET_LABEL}-
oc label node ovn-worker ${SOURCE_LABEL}-
oc label node ovn-worker2 ${SOURCE_LABEL}-
oc label node ovn-control-plane ${EGRESS_ASSIGNABLE_LABEL}- --overwrite
oc label node ovn-worker ${EGRESS_ASSIGNABLE_LABEL}- --overwrite
oc label node ovn-worker2  ${EGRESS_ASSIGNABLE_LABEL}- --overwrite
~~~

Run the E2E:
~~~
[root@ovnkubernetes ~]# cd development/ovn-kubernetes/test/e2e/
[root@ovnkubernetes e2e]# go test -v . --count=1 -ginkgo.focus "Should validate the egress IP SNAT functionality against host-networked pods" -ginkgo.v  
~~~

Test passes:
~~~
e2e egress IP validation
/root/development/ovn-kubernetes/test/e2e/e2e.go:811
  Should validate the egress IP SNAT functionality against host-networked pods
  /root/development/ovn-kubernetes/test/e2e/e2e.go:1204
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSS
Ran 1 of 67 Specs in 41.779 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 66 Skipped
~~~

==========================================================

### Appendix

[0]

01) separate nodes into 2 sets:
~~~
export SOURCE_NAMESPACE="egressip-source"
export TARGET_NAMESPACE="egressip-target"
export EGRESSIP_NAME="egressip-source"
export SOURCE_LABEL="node-role.kubernetes.io/egressip-test-source"
export TARGET_LABEL="node-role.kubernetes.io/egressip-test-target"
export TARGET_TAINT="egressip-test-target"
export EGRESS_ASSIGNABLE_LABEL="k8s.ovn.org/egress-assignable"
# https://docs.openshift.com/container-platform/4.10/installing/installing_aws/installing-restricted-networks-aws.html#installation-cloudformation-security_installing-restricted-networks-aws
export TARGET_PORT="32767"

oc label node ovn-control-plane $TARGET_LABEL=""
oc label node ovn-worker $SOURCE_LABEL=""
oc label node ovn-worker2 $SOURCE_LABEL=""
~~~

02) apply a taint on the target node
~~~
echo "Applying target NoExecute taint to target nodes"
nodes=$(oc get nodes -l ${TARGET_LABEL}= -o name)
if [ $(echo $nodes | wc -w) -lt 1 ] ; then
  echo "Not enough worker nodes with label ${TARGET_LABEL} - at least 1 worker node is required. Got: $nodes"
  exit 1
fi

i=0
for n in $nodes; do
  oc adm taint node $n ${TARGET_TAINT}=true:NoExecute --overwrite
done
~~~

03) assign labels to source nodes:
~~~
echo "Applying egress assignable label ${EGRESS_ASSIGNABLE_LABEL} to nodes"
nodes=$(oc get nodes -l ${SOURCE_LABEL}= -o name)
if [ $(echo $nodes | wc -w) -lt 2 ] ; then
  echo "Not enough worker nodes with label ${SOURCE_LABEL} - at least 2 worker nodes are required. Got: $nodes"
  exit 1
fi

i=0
for n in $nodes; do
  oc label $n ${EGRESS_ASSIGNABLE_LABEL}="" --overwrite
done

oc get nodes -l k8s.ovn.org/egress-assignable=""
~~~

04) create target pod (host networked):
~~~
file=$(mktemp)
cat <<EOF > ${file}
---
apiVersion: v1
kind: Namespace
metadata:
  name: ${TARGET_NAMESPACE}
  labels:
    env: ${TARGET_NAMESPACE}
---
apiVersion: apps/v1
kind: "DaemonSet"
metadata:
  labels:
    app: ${TARGET_NAMESPACE}-deployment
  name: ${TARGET_NAMESPACE}-deployment
  namespace: ${TARGET_NAMESPACE}
spec:
  selector:
    matchLabels:
      app: ${TARGET_NAMESPACE}-deployment
  template:
    metadata:
      labels:
        app: ${TARGET_NAMESPACE}-deployment
    spec:
      hostNetwork: true
      nodeSelector:
        ${TARGET_LABEL}: ""
      tolerations:
        - key: ${TARGET_TAINT}
          operator: Exists
      containers:
      - command:
        - "/agnhost"
        - "netexec"
        - "--http-port"
        - "${TARGET_PORT}"
        image: k8s.gcr.io/e2e-test-images/agnhost:2.33
        imagePullPolicy: IfNotPresent
        name: agnhost
EOF
oc apply -f ${file}
~~~

05) create pods inside the source namespace which will get the egressip later:
~~~
file=$(mktemp)
cat <<EOF >| $file
---
apiVersion: v1
kind: Namespace
metadata:
  name: ${SOURCE_NAMESPACE}
  labels:
    env: ${SOURCE_NAMESPACE}
---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: ${SOURCE_NAMESPACE}-deployment
  name: ${SOURCE_NAMESPACE}-deployment
  namespace: ${SOURCE_NAMESPACE}
spec:
  replicas: 2
  selector:
    matchLabels:
      app: ${SOURCE_NAMESPACE}-deployment
  template:
    metadata:
      labels:
        app: ${SOURCE_NAMESPACE}-deployment
    spec:
      nodeSelector:
        ${SOURCE_LABEL}: ""
      containers:
      - command:
        - "/agnhost"
        - "netexec"
        - "--http-port"
        - "8000"
          #- serve-hostname
        image: k8s.gcr.io/e2e-test-images/agnhost:2.33
        imagePullPolicy: IfNotPresent
        name: agnhost
EOF
cat ${file}
oc apply -f ${file}
~~~

andreaskaris commented:
When using an EgressIP object with 2 EgressIPs, when assigning them to both ovn-worker and ovn-worker2 ... then when deleting the egressip assignable label from one of the workers ... The EgressIP SNAT will be deleted, but we do *not* recreate the default SNAT subsequently.

=========================================

Use the baseline from my previous comment, and then try this:

~~~
[root@ovnkubernetes ~]# oc get nodes -l k8s.ovn.org/egress-assignable=""
NAME          STATUS   ROLES                  AGE   VERSION
ovn-worker    Ready    egressip-test-source   35m   v1.23.3
ovn-worker2   Ready    egressip-test-source   35m   v1.23.3
~~~

~~~
oot@ovnkubernetes ~]# oc get pods -o wide -A | grep egress
egressip-source      egressip-source-deployment-56d66c78dd-44fdl   1/1     Running   0          7m29s   10.244.2.4   ovn-worker2         <none>           <none>
egressip-source      egressip-source-deployment-56d66c78dd-9m4rb   1/1     Running   0          7m29s   10.244.1.9   ovn-worker          <none>           <none>
egressip-target      egressip-target-deployment-q4x6v              1/1     Running   0          7m36s   172.18.0.2   ovn-control-plane   <none>           <none>
~~~

~~~
file=$(mktemp)
cat <<EOF >| ${file}
apiVersion: k8s.ovn.org/v1
kind: EgressIP
metadata:
  name: ${EGRESSIP_NAME}
spec:
  egressIPs: [ "172.18.0.10", "172.18.0.11" ]
  namespaceSelector:
    matchLabels:
      env: ${SOURCE_NAMESPACE}
EOF
oc apply -f $file
~~~

~~~
[root@ovnkubernetes ~]# oc get egressip -o yaml | grep status: -A8
        f:status:
          .: {}
          f:items: {}
      manager: ovn-control-plane
      operation: Update
      time: "2022-04-27T10:22:47Z"
    name: egressip-source
    resourceVersion: "5237"
    uid: ab49e38e-ec44-497e-a225-8dff5a78bcfd
--
  status:
    items:
    - egressIP: 172.18.0.10
      node: ovn-worker
    - egressIP: 172.18.0.11
      node: ovn-worker2
kind: List
metadata:
  resourceVersion: ""
~~~

~~~
[root@ovnkubernetes ~]# oc exec -n $SOURCE_NAMESPACE egressip-source-deployment-56d66c78dd-9m4rb -- curl -s  172.18.0.2:$TARGET_PORT/clientip
[root@ovnkubernetes ~]# oc exec -n $SOURCE_NAMESPACE egressip-source-deployment-56d66c78dd-44fdl -- curl -s  172.18.0.2:$TARGET_PORT/clientip
172.18.0.11:53662[root@ovnkubernetes ~]# 
~~~

~~~
[root@ovnkubernetes ~]#  oc rsh -n ovn-kubernetes $(oc get pods -o name -n ovn-kubernetes -l name=ovnkube-master)
Defaulting container name to ovn-northd.
Use 'oc describe pod/ovnkube-master-85d5b4f6b9-5sjsq -n ovn-kubernetes' to see all of the containers in this pod.
sh-5.1# ovn-nbctl lr-nat-list GR_ovn-worker2
TYPE             EXTERNAL_IP        EXTERNAL_PORT    LOGICAL_IP            EXTERNAL_MAC         LOGICAL_PORT
snat             172.18.0.11                         10.244.2.4
snat             172.18.0.11                         10.244.1.9
sh-5.1# ovn-nbctl lr-nat-list GR_ovn-worker
TYPE             EXTERNAL_IP        EXTERNAL_PORT    LOGICAL_IP            EXTERNAL_MAC         LOGICAL_PORT
snat             172.18.0.10                         10.244.2.4
snat             172.18.0.10                         10.244.1.9
snat             172.18.0.3                          10.244.1.3
snat             172.18.0.3                          10.244.1.4
snat             172.18.0.3                          10.244.1.5
~~~

Remove the label from ovn-worker2:
~~~
[root@ovnkubernetes ~]# oc label node ovn-worker2 ${EGRESS_ASSIGNABLE_LABEL}- --overwrite
node/ovn-worker2 labeled
~~~

New status:
~~~
[root@ovnkubernetes ~]# oc get egressip -o yaml | grep status: -A8
        f:status:
          .: {}
          f:items: {}
      manager: ovn-control-plane
      operation: Update
      time: "2022-04-27T10:22:47Z"
    name: egressip-source
    resourceVersion: "5527"
    uid: ab49e38e-ec44-497e-a225-8dff5a78bcfd
--
  status:
    items:
    - egressIP: 172.18.0.10
      node: ovn-worker
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
~~~

And now it breaks:
~~~
[root@ovnkubernetes ~]# oc exec -n $SOURCE_NAMESPACE egressip-source-deployment-56d66c78dd-44fdl -- curl -s  172.18.0.2:$TARGET_PORT/clientip
^C
[root@ovnkubernetes ~]# oc exec -n $SOURCE_NAMESPACE egressip-source-deployment-56d66c78dd-9m4rb -- curl -s  172.18.0.2:$TARGET_PORT/clientip
172.18.0.10:51396[root@ovnkubernetes ~]# 
~~~

~~~
[root@ovnkubernetes ~]#  oc rsh -n ovn-kubernetes $(oc get pods -o name -n ovn-kubernetes -l name=ovnkube-master)
Defaulting container name to ovn-northd.
Use 'oc describe pod/ovnkube-master-85d5b4f6b9-5sjsq -n ovn-kubernetes' to see all of the containers in this pod.
sh-5.1# ovn-nbctl lr-nat-list GR_ovn-worker2
sh-5.1# ovn-nbctl lr-nat-list GR_ovn-worker
TYPE             EXTERNAL_IP        EXTERNAL_PORT    LOGICAL_IP            EXTERNAL_MAC         LOGICAL_PORT
snat             172.18.0.10                         10.244.2.4
snat             172.18.0.10                         10.244.1.9
snat             172.18.0.3                          10.244.1.3
snat             172.18.0.3                          10.244.1.4
snat             172.18.0.3                          10.244.1.5
~~~

tssurya commented on test/e2e/e2e.go:
```diff
@@ -1160,6 +1176,182 @@ spec:
+	   8. Remove the "k8s.ovn.org/egress-assignable" label from egress1Node
+	   9. Check that the status is of length one and that it is assigned to egress2Node
+	   10. Check connectivity from pod to an external "node" and verify that the srcIP is the expected egressIP
+	   11. Check connectivity from pod to another node (egress2Node) and verify that the srcIP is the expected nodeIP (known inconsistency compared to step 6 for SGW)
```
yes

tssurya commented:
@andreaskaris : can you share the logs from the ovnkube-master container for the last test you did? deleteEgressIPAssignments should get called when a change in egressIP happens as in when 11 goes away and it should call addPerPodSNAT again which should add the SNAT for 1.5 again. Having logs will help figure out why this didn't happen

tssurya commented:
oh well there we go! good find @andreaskaris : 
```
// deleteEgressIPAssignments performs a full egress IP setup deletion on a per
// (egress IP name - status) basis. The idea is thus to list the full content of
// the NB DB for that egress IP object and delete everything which match the
// status. We also need to update the podAssignment cache and finally re-add the
// external GW setup in case the pod still exists.
func (oc *Controller) deleteEgressIPAssignments(name string, statusesToRemove []egressipv1.EgressIPStatusItem) error {
	oc.eIPC.podAssignmentMutex.Lock()
	defer oc.eIPC.podAssignmentMutex.Unlock()
	for _, statusToRemove := range statusesToRemove {
		klog.V(5).Infof("Deleting pod egress IP status: %v for EgressIP: %s", statusToRemove, name)
		if err := oc.eIPC.deleteEgressIPStatusSetup(name, statusToRemove); err != nil {
			return err
		}
		for podKey, podStatus := range oc.eIPC.podAssignment {
			delete(podStatus.egressStatuses, statusToRemove)
			if len(podStatus.egressStatuses) == 0 {
				podNamespace, podName := getPodNamespaceAndNameFromKey(podKey)
				if err := oc.eIPC.addPerPodGRSNAT(podNamespace, podName, podStatus.podIPs); err != nil {
					return err
				}
				delete(oc.eIPC.podAssignment, podKey)
			}
		}
	}
	return nil
}
```
Turns out we do call `deleteEgressIPAssignments` which should add back the SNAT, BUT, Alex had a clause `if len(podStatus.egressStatuses) == 0 {` like only add back the egressGW setup if pod has 0 egressIPs on it which is not the case in your testing.

Actually this boils down to the same problem. Why didn't we explicitly state one pod can be managed only by one EIP when defining the feature? Because rn it could randomly pick any egressIP I guess. Anyways will update this to include this portion and add an e2e change to include this as well.

tssurya commented:
```
so it seems like re-assignment of egressIPs to other nodes is taking more than 40seconds?:
```
its not a timing issue at all, something else entirely, we have changed the log around to capture the patch update that's actually happening here which causes the flake.

andreaskaris commented:
I looked at the latest code changes and retested the previous failing https://github.com/ovn-org/ovn-kubernetes/pull/2945#issuecomment-1110840276 with the new code in shared GM and it works fine. I'll just have another look at it all in local gm mode before lgtm'ing.

andreaskaris commented:
/lgtm    (I did a quick sanity check with local gm and it looks ok)

flavio-fernandes reviewed (commented):
/lgtm
TY Surya!
